lib/rspec/expectations/failure_aggregator.rb

```diff
@@ -0,0 +1,159 @@
+module RSpec
+  module Expectations
+    # @private
+    class FailureAggregator
+      attr_reader :block_label, :metadata
+
+      def aggregate
+        RSpec::Support.with_failure_notifier(self) do
+          begin
+            yield
+          rescue ExpectationNotMetError => e
+            failures << e
+          rescue Exception => e
+            other_errors << e
+          end
+        end
+
+        raise_aggregated_failures
+      end
+
+      def failures
+        @failures ||= []
+      end
+
+      def other_errors
+        @other_errors ||= []
+      end
+
+      # This method is defined to satisfy the callable interface
+      # expected by `RSpec::Support.with_failure_notifier`.
+      def call(failure)
+        failure.set_backtrace(caller) unless failure.backtrace
+        failures << failure
+      end
+
+    private
+
+      def initialize(block_label, metadata)
+        @block_label = block_label
+        @metadata    = metadata
+      end
+
+      def raise_aggregated_failures
+        all_errors = failures + other_errors
+
+        case all_errors.size
+        when 0 then return nil
+        when 1 then raise all_errors.first
```

Comment:
I'm ok with this behaviour of only raise multiple if there are indeed more than one.